File: src/models/id.rs
use serde::{Deserialize, Serialize};
use std::fmt;
use uuid::Uuid;

/// Opaque identifier for stored entities.
///
/// For file-backed storage, ids should be safe path segments (no slashes).
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(transparent)]
pub struct Id(String);

impl Default for Id {
    fn default() -> Self {
        Self::new()
    }
}

impl Id {
    /// Namespace UUID for generating deterministic IDs from external identifiers.
    const NAMESPACE: Uuid = Uuid::from_u128(0x6ba7b810_9dad_11d1_80b4_00c04fd430c8);

    pub fn new() -> Self {
        Self(Uuid::new_v4().to_string())
    }

    /// Create an ID from an arbitrary string.
    /// Note: The string must be a valid path segment (no slashes).
    pub fn from_string(value: impl Into<String>) -> Self {
        Self(value.into())
    }

    /// Create a deterministic, filesystem-safe ID from an external identifier.
    /// Uses UUID5 to hash the input, ensuring the same input always produces the same ID.
    /// This is useful for external IDs that may contain special characters (like base64).
    pub fn from_external(value: &str) -> Self {
        Self(Uuid::new_v5(&Self::NAMESPACE, value.as_bytes()).to_string())
    }

    pub fn as_str(&self) -> &str {
        &self.0
    }
}

impl fmt::Display for Id {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.0.fmt(f)
    }
}

impl From<String> for Id {
    fn from(value: String) -> Self {
        Self(value)
    }
}

impl From<&str> for Id {
    fn from(value: &str) -> Self {
        Self(value.to_string())
    }
}

impl AsRef<str> for Id {
    fn as_ref(&self) -> &str {
        self.as_str()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_from_external_is_deterministic() {
        let first = Id::from_external("schwab-account-123");
        let second = Id::from_external("schwab-account-123");
        assert_eq!(first, second);
    }

    #[test]
    fn test_from_external_differs_for_different_inputs() {
        let first = Id::from_external("schwab-account-123");
        let second = Id::from_external("schwab-account-456");
        assert_ne!(first, second);
    }

    #[test]
    fn test_from_external_is_path_safe() {
        let id = Id::from_external("weird/account/value");
        assert!(!id.as_str().contains('/'));
    }

    #[test]
    fn test_from_string_keeps_value() {
        let id = Id::from_string("account-id-123");
        assert_eq!(id.as_str(), "account-id-123");
    }
}

File: ts/src/models/id.ts
import { v4 as uuidv4, v5 as uuidv5 } from 'uuid';

/**
 * Error thrown when an Id value fails path-safety validation.
 */
export class IdError extends Error {
  public readonly value: string;

  constructor(value: string) {
    super(
      `Invalid id ${JSON.stringify(value)}: ids must be a single path segment (no '/', '\\', NUL, '.' or '..')`,
    );
    this.name = 'IdError';
    this.value = value;
  }
}

/**
 * Opaque identifier for stored entities.
 *
 * For file-backed storage, ids should be safe path segments (no slashes).
 * Serializes transparently as a plain string in JSON.
 */
export class Id {
  /** Namespace UUID for generating deterministic IDs from external identifiers. */
  private static readonly NAMESPACE = '6ba7b810-9dad-11d1-80b4-00c04fd430c8';

  private readonly inner: string;

  private constructor(value: string) {
    this.inner = value;
  }

  /** Create a new random Id (UUID v4). */
  static new(): Id {
    return new Id(uuidv4());
  }

  /** Create an Id from an arbitrary string (no validation). */
  static fromString(value: string): Id {
    return new Id(value);
  }

  /**
   * Create an Id from a string, validating that it is a safe path segment.
   * Throws IdError if the value is not path-safe.
   */
  static fromStringChecked(value: string): Id {
    if (Id.isPathSafe(value)) {
      return new Id(value);
    }
    throw new IdError(value);
  }

  /**
   * Create a deterministic, filesystem-safe Id from an external identifier.
   * Uses UUID v5 with a fixed namespace, so the same input always produces the same Id.
   * Useful for external IDs that may contain special characters (like base64).
   */
  static fromExternal(value: string): Id {
    return new Id(uuidv5(value, Id.NAMESPACE));
  }

  /** Returns true if the string is safe to use as a single path segment. */
  static isPathSafe(value: string): boolean {
    if (value === '' || value === '.' || value === '..') {
      return false;
    }
    for (let i = 0; i < value.length; i++) {
      const ch = value[i];
      if (ch === '/' || ch === '\\' || ch === '\0') {
        return false;
      }
    }
    return true;
  }

  /** The underlying string value. */
  asStr(): string {
    return this.inner;
  }

  /** Value equality comparison. */
  equals(other: Id): boolean {
    return this.inner === other.inner;
  }

  /** String coercion returns the inner value. */
  toString(): string {
    return this.inner;
  }

  /** JSON serialization returns the plain string (transparent). */
  toJSON(): string {
    return this.inner;
  }
}
